eng/pipelines/installer/jobs/base-job.yml

```diff
@@ -0,0 +1,182 @@
+
+  - ${{ if eq(parameters.osGroup, 'Windows_NT') }}:
+    - name: setScriptToEchoAndFailOnNonZero
+      value: ''
+  - ${{ if ne(parameters.osGroup, 'Windows_NT') }}:
+    # Stop the script if any command exits nonzero, and display each command.
+    - name: setScriptToEchoAndFailOnNonZero
+      value: set -xe
+
+  # Produce test-signed build for PR and Public builds
+  - ${{ if or(eq(variables['System.TeamProject'], 'public'), in(variables['Build.Reason'], 'PullRequest')) }}:
+    - name: SignType
+      value: test
+
+  # Set up non-PR build from internal project
+  - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+    - name: SignType
+      value: $[ coalesce(variables.OfficialSignType, 'real') ]
+
+  - name: LiveOverridePathArgs
+    value: >-
+      $(CoreCLROverridePathArg)
+      $(CoreFXOverridePathArg)
+
+  - name: CoreCLROverridePathArg
+    value: ''
+  - name: CoreFXOverridePathArg
+    value: ''
+
+  - name: CoreClrDownloadPath
+    value: ''
+  - name: LibrariesDownloadPath
+    value: ''
+
+  - ${{ if ne(parameters.liveCoreClrBuildConfig, '') }}:
+    - name: liveCoreClrLegName
+      value: ${{ format('{0}{1}_{2}_{3}',
+        parameters.osGroup,
+        parameters.osSubgroup,
+        parameters.archType,
+        parameters.liveCoreClrBuildConfig) }}
+    - name: CoreClrDownloadPath
+      value: 'artifacts/transport/coreclr'
+    - name: CoreCLROverridePathArg
+      value: /p:CoreCLROverridePath=${{ parameters.buildCommandSourcesDirectory }}$(CoreClrDownloadPath)
+    - name: CoreClrArtifactName
+      value: CoreCLRProduct_$(liveCoreClrLegName)
+
+  - ${{ if ne(parameters.liveLibrariesBuildConfig, '') }}:
+    - name: liveLibrariesLegName
+      value: ${{ format('{0}{1}_{2}_{3}',
+        parameters.osGroup,
+        parameters.osSubgroup,
+        parameters.archType,
+        parameters.liveLibrariesBuildConfig) }}
+    - name: LibrariesDownloadPath
+      value: 'artifacts/transport/libraries'
+    - name: CoreFXOverridePathArg
+      value: /p:CoreFXOverridePath=${{ parameters.buildCommandSourcesDirectory }}$(LibrariesDownloadPath)
+    - name: LibrariesArtifactName
+      value: libraries_bin_$(liveLibrariesLegName)
+
+  dependsOn:
+  - checkout
+  - ${{ parameters.dependsOn }}
+  - ${{ if ne(parameters.liveCoreClrBuildConfig, '') }}:
+    - coreclr_product_build_${{ format('{0}{1}_{2}_{3}',
+        parameters.osGroup,
+        parameters.osSubgroup,
+        parameters.archType,
+        parameters.liveCoreClrBuildConfig) }}
+  - ${{ if ne(parameters.liveLibrariesBuildConfig, '') }}:
+    - libraries_build_netcoreapp_${{ format('{0}{1}_{2}_{3}',
+        parameters.osGroup,
+        parameters.osSubgroup,
+        parameters.archType,
+        parameters.liveLibrariesBuildConfig) }}
+
+  steps:
+
+  - ${{ parameters.prepareSteps }}
+
+  - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
```

Comment:
This could be moved to an `_IsOfficialBuild` common variable and we could add a variable template that is imported in runtime.yml that way we avoid duplication and these hard to read conditions. Something like: https://github.com/dotnet/runtime/blob/master/eng/pipelines/libraries/variables.yml#L9

Reply:
Interesting... I never got to grips with variable templates (or using variable values in template expressions), this is a good starting point. I'll give this a shot, thanks.

Reply:
Just please remember the variable templates can only be imported at the "top level" i.e. at the pipeline YAML level, not from the lower-level templates, that's the current limitation AFAIK. Like - it would be super-logical to import a variable template section into one of the CoreCLR template - it just doesn't work. I initially tried to mimic variables in templates by splitting platform-matrix in three different templates, then I rolled it back because it plain sucked. Santi currently worked it around by using explicit nested templates (the xplat-setup stuff). I believe this is just poor design on Azure Pipelines' part.

Reply:
Yeah... weird limits on templates. I don't see any fundamental reason you shouldn't be able to use them anywhere you want. (Whether it gets messy is up to the author!) Really confusing.

I guess this specific weirdness is because *pipeline* variables are "statically defined", but *job* variables aren't... but the only reference I find to statically defined variables is https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops so I can't tell if that's what's intended:

> In a parse time expression, you have access to `parameters` and statically-defined `variables`.

There's feedback filed on a few pages about this being confusing. 😛

Reply:
Yeah. It would have to be imported in runtime.yml and then you can do `variables['IsOfficialBuild']`.... That's what we're doing for the libraries build now.